metricbeat/mb/testing/testdata.go

```diff
@@ -0,0 +1,408 @@
+	return config
+}
+
+// TestDataConfig is a convenience helper function to read the testdata config
+// from the usual path
+func TestDataConfig(t *testing.T) DataConfig {
+	t.Helper()
+	return ReadDataConfig(t, "_meta/testdata/config.yml")
+}
+
+// TestDataFiles run tests from the usual path (`_meta/testdata`)
+func TestDataFiles(t *testing.T, module, metricSet string) {
+	t.Helper()
+	config := TestDataConfig(t)
+	TestDataFilesWithConfig(t, module, metricSet, config)
+}
+
+// TestDataFiles run tests for a testdata config
+func TestDataFilesWithConfig(t *testing.T, module, metricSet string, config DataConfig) {
+	t.Helper()
+	ff, err := filepath.Glob(filepath.Join(config.Path, "*."+config.Suffix))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(ff) == 0 {
+		t.Fatalf("test path with config but without data files: %s", config.Path)
+	}
+
+	var files []string
+	for _, f := range ff {
+		// Exclude all the expected files
+		if strings.HasSuffix(f, expectedExtension) {
+			continue
+		}
+		files = append(files, f)
+	}
+
+	for _, f := range files {
+		t.Run(filepath.Base(f), func(t *testing.T) {
+			runTest(t, f, module, metricSet, config)
+		})
+	}
+}
+
+func runTest(t *testing.T, file string, module, metricSetName string, config DataConfig) {
+	// starts a server serving the given file under the given url
+	s := server(t, file, config.URL)
+	defer s.Close()
+
+	moduleConfig := getConfig(module, metricSetName, s.URL, config)
+	metricSet := NewMetricSet(t, moduleConfig)
+
+	var events []mb.Event
+	var errs []error
+
+	switch v := metricSet.(type) {
+	case mb.ReportingMetricSetV2:
+		metricSet := NewReportingMetricSetV2(t, moduleConfig)
+		events, errs = ReportingFetchV2(metricSet)
+	case mb.ReportingMetricSetV2Error:
+		metricSet := NewReportingMetricSetV2Error(t, moduleConfig)
+		events, errs = ReportingFetchV2Error(metricSet)
+	default:
+		t.Fatalf("unknown type: %T", v)
+	}
+
+	// Gather errors to build also error events
+	for _, e := range errs {
+		// TODO: for errors strip out and standardise the URL error as it would create a different diff every time
+		events = append(events, mb.Event{Error: e})
+	}
+
+	var data []common.MapStr
+
+	for _, e := range events {
+		beatEvent := StandardizeEvent(metricSet, e, mb.AddMetricSetInfo)
+		// Overwrite service.address as the port changes every time
+		beatEvent.Fields.Put("service.address", "127.0.0.1:55555")
+		data = append(data, beatEvent.Fields)
+	}
+
+	// Sorting the events is necessary as events are not necessarily sent in the same order
+	sort.SliceStable(data, func(i, j int) bool {
+		h1, _ := hashstructure.Hash(data[i], nil)
+		h2, _ := hashstructure.Hash(data[j], nil)
+		return h1 < h2
+	})
+
+	if err := checkDocumented(t, data, config.OmitDocumentedFieldsCheck); err != nil {
+		t.Errorf("%v: check if fields are documented in `metricbeat/%s/%s/_meta/fields.yml` "+
+			"file or run 'make update' on Metricbeat folder to update fields in `metricbeat/fields.yml`",
+			err, module, metricSetName)
```

Comment:
@sayden I changed this line when merging with #11780 so module and metricset appear in the error message.
Before:
```
--- FAIL: TestData (0.02s)
    --- FAIL: TestData/docs.json (0.02s)
        testdata.go:209: 'field missing 'service.namexx'' check if fields are documented in `metricbeat/{module}/{metricset}/_meta/fields.yml` file or run 'make update' on Metricbeat folder to update root `metricbeat/fields.yml` in 
```
With this change:
```
--- FAIL: TestData (0.02s)
    --- FAIL: TestData/docs.json (0.02s)
        testdata.go:209: field missing 'service.namexx': check if fields are documented in `metricbeat/traefik/health/_meta/fields.yml` file or run 'make update' on Metricbeat folder to update fields in `metricbeat/fields.yml`
```

Reply:
Good idea in fact